src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/Serialization/NamingStyle.cs

```diff
@@ -0,0 +1,289 @@
+        public string WordSeparator { get; set; }
+        [DataMember]
+        public Capitalization CapitalizationScheme { get; set; }
+
+        public IEnumerable<string> InferWords(string name)
+        {
+            // TODO
+            return SpecializedCollections.SingletonEnumerable(name);
+        }
+
+        public string CreateName(IEnumerable<string> words)
+        {
+            EnsureNonNullProperties();
+
+            var wordsWithCasing = ApplyCapitalization(words);
+            var combinedWordsWithCasing = string.Join(WordSeparator, wordsWithCasing);
+            return Prefix + combinedWordsWithCasing + Suffix;
+        }
+
+        private IEnumerable<string> ApplyCapitalization(IEnumerable<string> words)
+        {
+            switch (CapitalizationScheme)
+            {
+                case Capitalization.PascalCase:
+                    return words.Select(w => CapitalizeFirstLetter(w));
+                case Capitalization.CamelCase:
+                    return words.Take(1).Select(w => DecapitalizeFirstLetter(w)).Concat(words.Skip(1).Select(w => CapitalizeFirstLetter(w)));
+                case Capitalization.FirstUpper:
+                    return words.Take(1).Select(w => CapitalizeFirstLetter(w)).Concat(words.Skip(1).Select(w => DecapitalizeFirstLetter(w)));
+                case Capitalization.AllUpper:
+                    return words.Select(w => w.ToUpper());
+                case Capitalization.AllLower:
+                    return words.Select(w => w.ToLower());
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private string CapitalizeFirstLetter(string word)
+        {
+            var chars = word.ToCharArray();
+            return new string(chars.Take(1).Select(c => char.ToUpper(c)).Concat(chars.Skip(1)).ToArray());
+        }
+
+        private string DecapitalizeFirstLetter(string word)
+        {
+            var chars = word.ToCharArray();
+            return new string(chars.Take(1).Select(c => char.ToLower(c)).Concat(chars.Skip(1)).ToArray());
+        }
+
+        public bool IsNameCompliant(string name, out string failureReason)
+        {
+            EnsureNonNullProperties();
+
+            if (!name.StartsWith(Prefix))
+            {
+                failureReason = string.Format(FeaturesResources.NamingViolation_MissingPrefix, Prefix);
+                return false;
+            }
+
+            if (!name.EndsWith(Suffix))
+            {
+                failureReason = string.Format(FeaturesResources.NamingViolation_MissingSuffix, Suffix);
+                return false;
+            }
+
+            if (name.Length <= Prefix.Length + Suffix.Length)
+            {
+                failureReason = string.Empty;
+                return true;
+            }
+
+            name = name.Substring(Prefix.Length);
+            name = name.Substring(0, name.Length - Suffix.Length);
+
+            var words = new[] { name };
```

Comment:
I'm (sadly) quite concerned about the amount of allocations caused by Linq in this method. I forget whether 4.6 fixed it so that arrays don't allocate in their enumerators.  If not, consider retyping this to ImmutableArray<string> to get our specialized enumerator and linq operators that work directly on it.